### PR TITLE
refactor route logic to reuse code

### PR DESCRIPTION
**Pull request by ramaraochavali**
carved out from https://github.com/istio/istio/pull/34247 to simplify review. No functional changes.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@hzxuzhonghu PTAL. Once this is merged, I will rebase https://github.com/istio/istio/pull/34247 so that we can review the merge logic.

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 1]
```diff
@@ -284,6 +284,38 @@ func BuildSidecarOutboundVirtualHosts(node *model.Proxy, push *model.PushContext
 	vhdomains := sets.Set{}
 	knownFQDN := sets.Set{}
 
+	buildVirtualHost := func(hostname string, vhwrapper istio_route.VirtualHostWrapper, svc *model.Service) *route.VirtualHost {
+		name := domainName(hostname, vhwrapper.Port)
+		if duplicateVirtualHost(name, vhosts) {
+			// This means this virtual host has caused duplicate virtual host name.
+			push.AddMetric(model.DuplicatedDomains, name, node.ID, fmt.Sprintf("duplicate domain from virtual service: %s", name))
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:291]: nit: before it was "from service" in the second case, here its always "virtual service"

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute.go]: Yeah. Good catch. Fixed now

**Pull request merged**